core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java

```diff
@@ -556,4 +558,118 @@ public void testFailAllReplicasInitializingOnPrimaryFailWhileHavingAReplicaToEle
         ShardRouting newPrimaryShard = clusterState.routingTable().index("test").shard(0).primaryShard();
         assertThat(newPrimaryShard, not(equalTo(primaryShardToFail)));
     }
+
+    public void testReplicaOnNewestVersionIsPromoted() {
```

Comment:
This test checks one specific scenario. I think that it can be easily generalized in the way of the `IndicesClusterStateServiceRandomUpdatesTests` so that it simulates a large range of scenarios.
Essentially it would boil down to creating a few nodes with random version (see `randomInitialClusterState` of `IndicesClusterStateServiceRandomUpdatesTests`), allocating a few shards to the nodes (see `ClusterStateChanges.createIndex`), then failing some of the shards (incl. primary), see `ClusterStateChanges.applyFailedShards` or failing some of the nodes (incl. primary), see `ClusterStateChanges.deassociateDeadNodes` and then checking that the new primary is on the node with highest version.